torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -347,6 +356,300 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+  /*
+  A utility function for returning a string representation of a std::vector containing ints/floats.
+  Useful for printing the shape of an N-dimensional tensor. This is used in the error message creation
+  part in the _extract_ndarrays function below.
+  */
+  std::ostringstream output;
+  output << "(";
+
+  if (end < 0) {
+    end = vec.size();
+  }
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << vec[i] << ", ";        
+    } else {
+      output << vec[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+template <typename T>
+std::string get_buffer_str(void* buffer, int64_t start, int64_t end) {
+  /*
+  A similar utility function as get_vector_str, except it works on an array buffer instead of vector.
+  */
+  T* array = static_cast<T*>(buffer); 
+  std::ostringstream output;
+  output << "(";
+
+  for (int64_t i = start; i < end; ++i) {
+    if (i != end - 1) {
+      output << array[i] << ", ";        
+    } else {
+      output << array[i] << ")";
+    }
+  }
+  return output.str();
+}  
+
+inline bool _shape_match_ndarray(std::vector<int64_t> & shape, PyArrayObject* array, int64_t start) {
+  /*
+  A utility function that returns true only if shape[start:] == array.shape where array is a numpy array.
+  Else returns false. This is used in _extract_ndarrays() below for checking whether an embedded numpy array
+  has the target shape.
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  if ((shape.size() - start) != ndim) {
+    return false;
+  }
+
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    if (array_shape[i] != shape[i+start]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void _populate_shape_vector(std::vector<int64_t> & shape, PyArrayObject* array) {
+  /*
+  A utility function for filling up the shape vector with entries from the shape array of the numpy array. 
+  */
+  int64_t ndim = PyArray_NDIM(array);
+  npy_intp* array_shape = PyArray_DIMS(array);
+  for (int64_t i = 0; i < ndim; ++i) {
+    shape.emplace_back(array_shape[i]);
+  }
+}
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<int64_t> & ndarray_shape, int64_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
+  /*
+  This is a helper function for extracting all the numpy arrays embedded within PySequences (recursively) representd by obj. This is done recursively in a 
+  depth first search style. It puts the numpy array pointers in the array_ptr_storage for later use. In case there is a non numpy array type within obj, 
+  this function immediately returns false. If only all the items inside obj are just numpy arrays, it returns true. It throws a ValueError in case of a dimension
+  mismatch. This function is used in extract_ndarrays() function below.
+  */
+  if (PyArray_Check(obj)) {
+    if (PyArray_NDIM((PyArrayObject*) obj) == 0) {
+      return false;
```

Comment:
This seems to violate the contract stipulated in the comment that it returns true if all the objects are NumPy arrays?

Reply:
I must have forgotten to update the comment.